mm2src/coins/eth/v2_activation.rs

```diff
@@ -114,50 +137,112 @@ pub struct EthNode {
+    InvalidPayload(String),
+    Transport(String),
 }
 
-impl From<AbortedError> for Erc20TokenActivationError {
-    fn from(e: AbortedError) -> Self { Erc20TokenActivationError::InternalError(e.to_string()) }
+impl From<AbortedError> for EthTokenActivationError {
+    fn from(e: AbortedError) -> Self { EthTokenActivationError::InternalError(e.to_string()) }
 }
 
-impl From<MyAddressError> for Erc20TokenActivationError {
+impl From<MyAddressError> for EthTokenActivationError {
     fn from(err: MyAddressError) -> Self { Self::InternalError(err.to_string()) }
 }
 
+impl From<GetNftInfoError> for EthTokenActivationError {
+    fn from(e: GetNftInfoError) -> Self {
+        match e {
+            GetNftInfoError::InvalidRequest(err) | GetNftInfoError::InvalidResponse(err) => {
+                EthTokenActivationError::InvalidPayload(err)
+            },
+            GetNftInfoError::ContractTypeIsNull => EthTokenActivationError::InvalidPayload(
+                "The contract type is required and should not be null.".to_string(),
+            ),
+            GetNftInfoError::Transport(err) => EthTokenActivationError::Transport(err),
+            GetNftInfoError::Internal(err) | GetNftInfoError::DbError(err) | GetNftInfoError::NumConversError(err) => {
+                EthTokenActivationError::InternalError(err)
+            },
+            GetNftInfoError::GetEthAddressError(err) => EthTokenActivationError::InternalError(err.to_string()),
+            GetNftInfoError::ParseRfc3339Err(err) => EthTokenActivationError::InternalError(err.to_string()),
+            GetNftInfoError::ProtectFromSpamError(err) => EthTokenActivationError::InternalError(err.to_string()),
+            GetNftInfoError::TransferConfirmationsError(err) => EthTokenActivationError::InternalError(err.to_string()),
+            GetNftInfoError::TokenNotFoundInWallet {
+                token_address,
+                token_id,
+            } => EthTokenActivationError::InternalError(format!(
+                "Token not found in wallet: {}, {}",
+                token_address, token_id
+            )),
+        }
+    }
+}
+
+impl From<ParseChainTypeError> for EthTokenActivationError {
+    fn from(e: ParseChainTypeError) -> Self { EthTokenActivationError::InternalError(e.to_string()) }
+}
+
+/// Represents the parameters required for activating either an ERC-20 token or an NFT on the Ethereum platform.
+#[derive(Clone, Deserialize)]
+#[serde(tag = "type", rename_all = "UPPERCASE")]
+pub enum EthTokenActivationParams {
+    Erc20(Erc20TokenActivationRequest),
+    Nft(NftActivationRequest),
+}
+
+/// Parameters for activating an ERC-20 token.
 #[derive(Clone, Deserialize)]
 pub struct Erc20TokenActivationRequest {
     pub required_confirmations: Option<u64>,
 }
 
+/// Parameters for activating an NFT.
+#[derive(Clone, Deserialize)]
+pub struct NftActivationRequest {
+    /// This URL is used to fetch necessary NFT details for activation
+    pub url: Url,
```

Comment:
We just simply set an url assuming it is moralis nft provider. I guess there may be other providers. Should we not extend this struct to indicate which provider is used?
Also do we have a unified approach how to configure providers in the mm2 (maybe we should always access providers via a proxy on mm2 nodes) @shamardy?

Reply:
> I guess there may be other providers.

nft feature currently supports only moralis V2 api formats. If you provide another nft api (not moralis), komodefi just wont be able to use it. Or what do you mean by other providers?

Reply:
I mean what if we want to support a different or yet another provider. Maybe we could indicate a provider in request, like:
```
pub struct NftActivationRequest {
    pub provider: NftProviderEnum,
    /// This URL is used to fetch necessary NFT details for activation
    pub url: Url,
}
```
Then we could use different providers or at least will know which provider GUI wants to use

Reply:
PS: I see they added [2.2 version](https://docs.moralis.io/web3-data-api/evm/reference/get-wallet-nft-transfers), need to check it. May be they did some optimizations.

Reply:
Still wondering, if the api expects just the moralis url why pass it in rpc and not just store inside the api?

Reply:
> Still wondering, if the api expects just the moralis url why pass it in rpc and not just store inside the api?

Keeping proxy url inside api means asking user to provide it in mm config at the start of application.
Well we can do it, so on the backend each time we will be searching nft_url field from `mm.conf` and deserializing it into `Url` type.

But lets take an example.
If we a talking about not our GUI team, but CLI users or other devs who is gona use komodefi, if they want to use different nft proxies for different reqs, they have to restart app with different config each time.

Then to reach a compromise, I would suggest unify both approach. Lets keep asking url to fetch nft info in requests, but have this param as an optional. And allow user to set default url in `mm.conf`. Thus, if url in request is `Some`, we dont use default url from config. If url from req is `None`, then we try to access it from config. If we dont find it or have error during deserialization string into `Url` type we propagate error.

@shamardy @dimxy what do you think

Reply:
> We are discussing how to connect to a gas api provider: it is going to be also via our proxy on mm2 nodes.
I guess for api users it is simpler when the api hides such details internally (that is, contains whichever used proxy in a config)

Could you please clarify where I can get acquainted with this information? I it in dm or on your PR or in feature issue?

I suggest to wait Omer's decision, also we can discuss API format next sprint.

Reply:
> Could you please clarify where I can get acquainted with this information? I it in dm or on your PR or in feature issue?

I am new to connecting api to providers. This is what I was told: connect via a proxy running on our nodes. So I guess this is a generic way how to connect to providers. With that in mind it seems logical for me if we store proxy urls in the config.
But yes, let's see how Omer sees this and we may do this identically both for nft and gas api providers

Reply:
> Also do we have a unified approach how to configure providers in the mm2 (maybe we should always access providers via a proxy on mm2 nodes) @shamardy?

We can consider providers similar to electrums/web3 nodes where we specify them in the activation request, we also have gas station url in eth which is treated similarly
https://github.com/KomodoPlatform/komodo-defi-framework/blob/81af437ffdc071959c7727ff77d67dbc71570e55/mm2src/coins/eth/v2_activation.rs#L119

> I mean what if we want to support a different or yet another provider. Maybe we could indicate a provider in request, like:
> 
> ```rust
> pub struct NftActivationRequest {
>     pub provider: NftProviderEnum,
>     /// This URL is used to fetch necessary NFT details for activation
>     pub url: Url,
> }
> ```
> Then we could use different providers or at least will know which provider GUI wants to use

I agree to this but we should enclose the provider information (e.g. url) inside the enum variants. We can use
```rust
#[serde(tag = "type", content = "info")]
```
so NftProviderEnum should be
```rust
#[derive(Deserialize)]
#[serde(tag = "type", content = "info")]
enum NftProviderEnum {
    Moralis {
        url: Url,
    }
    // other providers
}
```
and we can set a default url if we need by implementing default for NftProviderEnum. But I don't think it's needed.

What do you think @laruh @dimxy ?

Reply:
> we can set a default url if we need by implementing default for NftProviderEnum

I dont get how are you going to provide default url by implementation `Default` for `NftProviderEnum`. Do you want to hard code it?

Also I dont get why do we need two urls in req structure (if we want to store url in NftProviderEnum and in url parameter)
```rust
pub struct NftActivationRequest {
    pub provider: NftProviderEnum,
    /// This URL is used to fetch necessary NFT details for activation
    pub url: Url,
}
```

If you suggest to have only enum type in req, then again do you want to hard code default url? 
```rust
pub struct NftActivationRequest {
    pub provider: NftProviderEnum
}
```

Right now we utilize only one url: "https://moralis-proxy.komodo.earth"

Reply:
about the default url, I think we can store it in MM2.json but allow to replace in the request (with optional url param)

Reply:
> in my fee per gas estimation PR I implemented this struct

Got it, thank you!
According to your example (I fixed a bit my code in [this my comment](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1501834040), removed url from Moralis struct in NftProviderEnum) I suppose right now we have almost the same suggestion related to default url implementation.

In addition, I suggested another variant of default url handling for NFTs to Omer via dm.
Instead of keeping nft url in `mm.config`, we can always ask `default_provider`, when we enable nft

```rust
pub struct NftActivationRequest {
    pub provider: NftProviderEnum,
}

#[derive(Deserialize)]
enum NftProviderEnum {
    Moralis {
        url: Url,
    },
    Other_provider
        url: Url,
    },
}
```

and add `nft_provider: Option<NftProviderEnum>` in `EthCoinImpl` structure. For eth coins it will be None, for global NFT it will be Some, bcz during activation user needs to provide provider type (which will be treated as default) to fetch nft data and activate nft. So when user calls `update_nft` or `transfer history` requests with None in url, we need to find activated global NFT and access it's default url provider, which must be Some. So this solution avoids keeping nft url in `mm.config`.

Reply:
In reply to this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1501824427 by @laruh

> I dont get how are you going to provide default url by implementation Default for NftProviderEnum. Do you want to hard code it?

Yep, that's what I meant but as I said we probably don't need it.

> Also I dont get why do we need two urls in req structure (if we want to store url in NftProviderEnum and in url parameter)

Well, I didn't suggest that, I was quoting your comment :)

In reply to this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1502118428 by @dimxy

> in my fee per gas estimation PR I implemented this struct
(it is stored in MM2.json)

`MM2.json` shouldn't have coin specific config. As I see it, we have 4 different levels of setting configs:

1. Hardcoded config: This is for setting a config that is that stays the same for all kind of users running mm2.
2. mm2 config or mm2.json: this is to set global variables related to mm2 instance before running, it can change between instances.
3. coin config: this is for settings that stay the same for the coin across multiple instances of mm2.
4. coin activation requests: this changes with each time we enable/disable a coin.

Each level can be overridden by any level below it. E.g. you can override coin config by activation requests. So if we need to place such default configs for gas provider url or nft provider url, they can be placed in coin configs. But I don't see the need for it and we can just include it each time in the request. I will leave it up to you @laruh @dimxy  to decide this though :)

Reply:
P.S. I suggested hardcoded config @laruh because we will not include NFT coin in coins config.

Reply:
>  But I don't see the need for it and we can just include it each time in the request.

Thanks for the clarification. Yes, I have the same opinion, that there is no problem with providing necessary url in each request. Pretty simple and flexible api, if user wants to have same url for each api let it be, all req have different urls - no problem. and no optional url params in requests.

Reply:
As suggested here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049/files#discussion_r1501097457

Please use the below since it will be easy to extend in the future without breaking backward compatibility unlike using url only.
```rust
pub struct NftActivationRequest {
    pub provider: NftProviderEnum
}
```
```rust
#[derive(Deserialize)]
#[serde(tag = "type", content = "info")]
enum NftProviderEnum {
    Moralis {
        url: Url,
    }
    // placeholder other providers
}
```